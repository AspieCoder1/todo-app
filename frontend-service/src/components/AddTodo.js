import React, { Component } from 'react';
import classNames from 'classnames';

import styles from '../styles/components/addTodo.module.scss';
export default class AddTodo extends Component {
	render() {
		return (
			<form onSubmit={this.props.onSubmit}>
				<div className={styles.inputGroup}>
					<input
						className={classNames(styles.input, {
							[styles.inputError]:
								this.props.errors && this.props.errors.noTodo,
						})}
						type="text"
						name="todoToAdd"
						value={this.props.todoToAdd}
						onChange={this.props.onChange}
						placeholder="Enter your new todo here..."
					/>
					<button className={styles.button} type="submit">
						Add Todo
					</button>
				</div>
				{this.props.errors && this.props.errors.noTodo ? (
					<p className={styles.error}>Please provide a todo</p>
				) : null}
			</form>
		);
	}
}
