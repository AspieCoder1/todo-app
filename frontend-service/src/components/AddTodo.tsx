import * as React from 'react';
import classNames from 'classnames';
import * as styles from '../styles/components/addTodo.module.scss';

interface IErrors {
	[key: string]: string;
}

interface IProps {
	errors: IErrors | undefined;
	todoToAdd: string;
	onSubmit: (e: React.FormEvent<HTMLFormElement>) => void;
	onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const AddTodo: React.FC<IProps> = props => (
	<form onSubmit={props.onSubmit}>
		<div className={styles.inputGroup}>
			<input
				className={classNames(styles.input, {
					[styles.inputError]: props.errors && props.errors.noTodo,
				})}
				type="text"
				name="todoToAdd"
				value={props.todoToAdd}
				onChange={props.onChange}
				placeholder="Enter your new todo here..."
			/>
			<button className={styles.button} type="submit">
				Add Todo
			</button>
		</div>
		{props.errors && props.errors.noTodo ? (
			<p className={styles.error}>Please provide a todo</p>
		) : null}
	</form>
);

export default AddTodo;
