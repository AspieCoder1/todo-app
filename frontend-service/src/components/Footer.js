import React from 'react';
import moment from 'moment';

export default () => {
	return (
		<div className="footer">
			<h3>&copy;AspieCoder {moment().year()}, all rights reserved</h3>
		</div>
	);
};
