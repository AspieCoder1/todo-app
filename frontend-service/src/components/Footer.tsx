import * as React from 'react';
import moment from 'moment';

const Footer: React.FC<{}> = () => {
	return (
		<div className="footer">
			<h3>&copy;AspieCoder {moment().year()}, all rights reserved</h3>
		</div>
	);
};

export default Footer;
