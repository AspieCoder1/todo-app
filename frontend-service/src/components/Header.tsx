import React from 'react';
import headerStyles from '../styles/components/header.module.scss';

const Header: React.FC<{}> = () => (
	<div>
		<div>
			<h2 className={headerStyles.title}>
				TodoApp - Manage your Todos in peace
			</h2>
		</div>
	</div>
);

export default Header;
