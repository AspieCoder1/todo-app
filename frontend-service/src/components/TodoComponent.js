import React from 'react';

import styles from '../styles/components/todoComponent.module.scss';

export default class TodoComponent extends React.Component {
	state = {
		completed: false,
	};

	componentDidMount() {
		this.setState({
			completed: this.props.todo.completed,
		});
	}

	render() {
		const { todo, removeTodo, onToggle } = this.props;
		return (
			<div className={styles.todo}>
				<input
					type="checkbox"
					name="completed"
					id="completed"
					checked={this.state.completed}
					value={this.state.completed}
					onChange={e => {
						this.setState({
							completed: e.target.checked,
						});

						onToggle(todo);
					}}
				/>
				<h3 className={styles.body}>{todo.body}</h3>
				<button className={styles.button} onClick={() => removeTodo(todo._id)}>
					Delete Todo
				</button>
			</div>
		);
	}
}
