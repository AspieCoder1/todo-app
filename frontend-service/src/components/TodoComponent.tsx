import * as React from 'react';

import styles from '../styles/components/todoComponent.module.scss';
import { ITodo } from '../../../backend-service/src/models/todo';
import { AxiosPromise, AxiosRequestConfig } from 'axios';

interface IProps {
	todo: ITodo;
	removeTodo: (
		todo: ITodo,
	) => (dispatch: any) => AxiosPromise<AxiosRequestConfig>;
	onToggle: (
		todo: ITodo,
	) => (dispatch: any) => AxiosPromise<AxiosRequestConfig>;
}

interface IState {
	completed: boolean;
}

export default class TodoComponent extends React.Component<IProps, IState> {
	state = {
		completed: false,
	};

	componentDidMount() {
		this.setState({
			completed: this.props.todo.completed,
		});
	}

	onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		const { todo, onToggle } = this.props;
		this.setState({ completed: e.target.checked });
		onToggle(todo);
	};

	render() {
		const { todo, removeTodo, onToggle } = this.props;
		return (
			<div className={styles.todo}>
				<input
					type="checkbox"
					name="completed"
					id="completed"
					checked={this.state.completed}
					// value={this.state.completed.toString()}
					onChange={this.onChange}
				/>
				<h3 className={styles.body}>{todo.body}</h3>
				<button className={styles.button} onClick={() => removeTodo(todo._id)}>
					Delete Todo
				</button>
			</div>
		);
	}
}
