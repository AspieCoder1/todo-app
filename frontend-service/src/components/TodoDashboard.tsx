import * as React from 'react';

import AddTodo from './AddTodo';
import Todos from './Todos';
import {
	addTodo,
	getTodos,
	removeTodo,
	editTodo,
} from '../redux/actions/todos';
import { logoutUser } from '../redux/actions/auth';
import { connect } from 'react-redux';
import Header from './Header';

import styles from '../styles/components/todoDashboard.module.scss';
import container from '../styles/components/container.module.scss';
import { AxiosPromise, AxiosRequestConfig } from 'axios';
import { bindActionCreators } from 'redux';
import { ITodo } from '../interfaces/todo';
import { ITodoReducer } from '../interfaces/reducer';

interface ITodoUpdates {
	body: string;
}

interface IProps extends IDispatchProps {
	todo: ITodoReducer;
	errors: any;
}

interface IDispatchProps {
	getTodos: () => (dispatch: any) => AxiosPromise<AxiosRequestConfig>;
	editTodo: (
		id: ITodo['_id'],
		todoData: any,
	) => (dispatch: any) => AxiosPromise<AxiosRequestConfig>;
	logoutUser: () => (dispatch: any) => void;
	addTodo: (
		todo: ITodoToAdd,
	) => (dispatch: any) => AxiosPromise<AxiosRequestConfig>;
	removeTodo: (
		id: ITodo['_id'],
	) => (dispatch: any) => AxiosPromise<AxiosRequestConfig>;
}

interface ITodoToAdd {
	todo: string;
}

interface IState {
	todos: ITodo[];
	filterCompleted: boolean;
	filteredTodos: ITodo[];
	todoToAdd: string;
}

export class TodoDashboard extends React.Component<IProps, IState> {
	state = {
		todos: [],
		filterCompleted: false,
		filteredTodos: [],
		todoToAdd: '',
	};

	componentDidMount() {
		this.props.getTodos();
	}

	onToggle = (todo: ITodo) => {
		this.editTodo(todo._id, {
			completed: !todo.completed,
		});
	};

	editTodo = (id: ITodo['_id'], todoData: any) => {
		this.props.editTodo(id, todoData);
	};

	static getDerivedStateFromProps(props: IProps) {
		return {
			todos: props.todo.todos,
			filteredTodos: props.todo.todos,
		};
	}

	removeTodo = (id: ITodo['_id']) => {
		if (this.state.todos.length > 0) {
			this.props.removeTodo(id);
			this.props.getTodos();
		}
	};

	onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		this.setState({ todoToAdd: e.target.value });
	};

	addTodo = (e: React.FormEvent<HTMLFormElement>) => {
		e.preventDefault();
		this.props.addTodo({ todo: this.state.todoToAdd });
		if (this.state.todoToAdd !== '') {
			this.props.getTodos();
		}
	};

	handleLogout = () => {
		this.props.logoutUser();
	};

	render() {
		const { loading } = this.props.todo;
		let content;
		if (loading) {
			content = <p>Loading ...</p>;
		} else {
			content = (
				<Todos
					todos={this.state.filteredTodos}
					removeTodo={this.removeTodo}
					onToggle={this.onToggle}
				/>
			);
		}
		return (
			<div className={styles.content}>
				<Header />
				<div className={container.wrapper}>
					<div>
						<h1>Your Todos</h1>
						<button onClick={this.handleLogout}>Logout</button>
					</div>
					<AddTodo
						onSubmit={this.addTodo}
						onChange={this.onChange}
						todoToAdd={this.state.todoToAdd}
						errors={this.props.errors}
					/>
					{content}
				</div>
			</div>
		);
	}
}

const mapStateToProps = (state: any) => ({
	todo: state.todos,
	errors: state.errors,
});

const mapDispatchToProps = (dispatch: any): IDispatchProps => {
	return bindActionCreators(
		{
			addTodo,
			getTodos,
			removeTodo,
			editTodo,
			logoutUser,
		},
		dispatch,
	);
};

export default connect(mapStateToProps, mapDispatchToProps)(TodoDashboard);
