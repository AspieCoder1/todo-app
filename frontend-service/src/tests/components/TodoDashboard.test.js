import React from 'react';
import { shallow, mount } from 'enzyme';

import { TodoDashboard } from '../../components/TodoDashboard';
import AddTodo from '../../components/AddTodo';
import TodoComponent from '../../components/TodoComponent';

it('should render TodoDashboard correctly', () => {
	const wrapper = shallow(
		<TodoDashboard todo={{ todos: [] }} getTodos={() => undefined} />,
	);

	expect(wrapper).toMatchSnapshot();
});

it('should call getTodos when component did mount', () => {
	const getTodos = jest.fn();
	shallow(<TodoDashboard todo={{ todos: [] }} getTodos={getTodos} />);

	expect(getTodos).toHaveBeenCalledTimes(1);
});

it('should update state when new todo is added is updated', () => {
	const wrapper = mount(
		<TodoDashboard todo={{ todos: [] }} getTodos={() => []} />,
	);

	wrapper
		.find(AddTodo)
		.find('input')
		.simulate('change', {
			target: {
				name: 'todoToAdd',
				value: 'Test',
			},
		});

	expect(wrapper.state().todoToAdd).toBe('Test');
});

it('should call onSubmit correctly', () => {
	const addTodo = jest.fn();
	const wrapper = mount(
		<TodoDashboard
			todo={{ todos: [] }}
			getTodos={() => []}
			addTodo={addTodo}
		/>,
	);

	wrapper
		.find(AddTodo)
		.find('input')
		.simulate('change', {
			target: {
				name: 'todoToAdd',
				value: 'Test',
			},
		});

	wrapper
		.find(AddTodo)
		.find('form')
		.simulate('submit');

	expect(addTodo).toHaveBeenCalledTimes(1);
	expect(addTodo).toHaveBeenCalledWith({ todo: 'Test' });
});

it('should handle logout correctly', () => {
	const logoutUser = jest.fn();
	const wrapper = shallow(
		<TodoDashboard
			todo={{ todos: [] }}
			getTodos={() => []}
			logoutUser={logoutUser}
		/>,
	);

	wrapper.find('button').simulate('click');

	expect(logoutUser).toHaveBeenCalledTimes(1);
});

it('should handle toggling todo correctly', () => {
	const editTodo = jest.fn();
	const wrapper = mount(
		<TodoDashboard
			todo={{ todos: [{ _id: 1, body: 'Test', completed: false }] }}
			editTodo={editTodo}
			getTodos={() => undefined}
		/>,
	);

	wrapper.find('#completed').simulate('change', {
		target: {
			checked: true,
		},
	});

	expect(editTodo).toHaveBeenCalledTimes(1);
});

it('should handle removeTodo correctly', () => {
	const removeTodo = jest.fn();
	const getTodos = jest.fn();
	const wrapper = mount(
		<TodoDashboard
			todo={{ todos: [{ _id: 1, body: 'Test', completed: false }] }}
			removeTodo={removeTodo}
			getTodos={getTodos}
		/>,
	);

	wrapper
		.find(TodoComponent)
		.find('button')
		.simulate('click');

	expect(removeTodo).toHaveBeenCalledTimes(1);
	expect(getTodos).toHaveBeenCalledTimes(2);
});
