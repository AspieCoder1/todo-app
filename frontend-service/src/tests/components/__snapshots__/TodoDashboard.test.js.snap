// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TodoDashboard correctly 1`] = `
<div
  className="content"
>
  <Header />
  <div
    className="wrapper"
  >
    <div>
      <h1>
        Your Todos
      </h1>
      <button
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
    <AddTodo
      onChange={[Function]}
      onSubmit={[Function]}
      todoToAdd=""
    />
    <_default
      onToggle={[Function]}
      removeTodo={[Function]}
      todos={Array []}
    />
  </div>
</div>
`;
